url2f44efcfa6_wangyudong_sohutw_java_sdk_2_0_tgz-pJ8-com_sohu_t_api_sdk_json_TestJ8/src/com/sohu/t/api/sdk/http/HttpClient.java

```diff
@@ -0,0 +1,961 @@
+
+        } else if ("POST".equals(method)) {
+
+            UTF8PostMethod post = new UTF8PostMethod(url);
+            for (PostParameter p : params) {
+                post.addParameter(p.getName(), p.getValue());
+            }
+            if (authenticated) {
+                post.addRequestHeader("Authorization", "OAuth2 " + this.oauth2AccessToken);
+            }
+
+            try {
+                httpclient.executeMethod(post);
+                response = new Response(post);
+            } catch (HttpException e) {
+                throw new TBlogException(e.getMessage(), e);
+            } catch (IOException e) {
+                throw new TBlogException(e.getMessage(), e);
+            }
+        }
+        return response;
+    }
+
+    public Response httpRequest(String url, PostParameter[] postParams, boolean authenticated, String httpMethod) throws TBlogException {
+        int retriedCount;
+        int retry = retryCount + 1;
+        Response res = null;
+        for (retriedCount = 0; retriedCount < retry; retriedCount++) {
+            int responseCode = -1;
+            try {
+                HttpURLConnection con = null;
+                OutputStream osw = null;
+                try {
+                    con = getConnection(url);
+                    con.setDoInput(true);
+                    setHeaders(url, postParams, con, authenticated, httpMethod);
+                    if (null != postParams || "POST".equals(httpMethod)) {
+                        con.setRequestMethod("POST");
+                        con.setRequestProperty("Content-Type",
+                                "application/x-www-form-urlencoded");
+                        con.setDoOutput(true);
+                        String postParam = "";
+                        if (postParams != null) {
+                            postParam = encodeParameters(postParams);
+                        }
+                        log("Post Params: ", postParam);
+                        byte[] bytes = postParam.getBytes("UTF-8");
+
+                        con.setRequestProperty("Content-Length",
+                                Integer.toString(bytes.length));
+                        osw = con.getOutputStream();
+                        osw.write(bytes);
+                        osw.flush();
+                        osw.close();
+                    } else if ("DELETE".equals(httpMethod)) {
+                        con.setRequestMethod("DELETE");
+                    } else {
+                        con.setRequestMethod("GET");
+                    }
+                    res = new Response(con);
+                    responseCode = con.getResponseCode();
+                    if (DEBUG) {
+                        log("Response: ");
+                        Map<String, List<String>> responseHeaders = con.getHeaderFields();
+                        for (String key : responseHeaders.keySet()) {
+                            List<String> values = responseHeaders.get(key);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.